pyogrio/_io.pyx

```diff
@@ -309,6 +309,42 @@ cdef get_driver(OGRDataSourceH ogr_dataset):
     return driver
 
 
+cdef get_feature_count(OGRLayerH ogr_layer):
+    """Get the feature count of a layer.
+
+    If GDAL returns an unknown count (-1), this iterates over every feature
+    to calculate the count.
+
+    Parameters
+    ----------
+    ogr_layer : pointer to open OGR layer
+
+    Returns
+    -------
+    int
+        count of features
+    """
+
+    cdef int feature_count = OGR_L_GetFeatureCount(ogr_layer, 1)
+
+    # if GDAL refuses to give us the feature count, we have to loop over all
+    # features ourselves and get the count.  This can happen for some drivers
+    # (e.g., OSM) or if a where clause is invalid but not rejected as error
+    if feature_count == -1:
+        # make sure layer is read from beginning
+        OGR_L_ResetReading(ogr_layer)
+
+        feature_count = 0
+        while True:
+            # No more rows available, so stop reading
+            if OGR_L_GetNextFeature(ogr_layer) == NULL:
+                break
+
+            feature_count +=1
```

Comment:
Only a matter of preference, but this could probably be written a bit more concise like this:
```suggestion
        while OGR_L_GetNextFeature(ogr_layer) != NULL:
            feature_count += 1
```